lib/pbench/server/api/resources/__init__.py

```diff
@@ -0,0 +1,861 @@
+    JSON client payload intake and validation.
+
+    Hooks are defined for subclasses extending this class to handle GET, POST,
+    PUT, and DELETE HTTP operations by overriding the abstract _get, _post,
+    _put, and _delete methods.
+    """
+
+    # We treat the current "access category" of the dataset as user-accessible
+    # metadata for the purposes of these APIs even though it's represented as
+    # a column on the Dataset model.
+    METADATA = Metadata.USER_METADATA + ["access", "owner"]
+
+    def __init__(
+        self,
+        config: PbenchServerConfig,
+        logger: Logger,
+        schema: Schema,
+        *,  # following parameters are keyword-only
+        role: API_OPERATION = API_OPERATION.READ,
+    ):
+        """
+        Base class constructor.
+
+        Args:
+            config: server configuration
+            logger: logger object
+            schema: API schema: for example,
+                    Schema(
+                        Parameter("user", ParamType.USER, required=True),
+                        Parameter("start", ParamType.DATE)
+                    )
+            role: specify the API role, defaulting to READ
+
+        NOTE: each class currently only supports a single schema across POST
+        and PUT operations. GET (and DELETE?) are assumed not to have/need a
+        request payload. If we ever need to change this, we can add a level
+        and describe a distinct Schema for each HTTP method.
+        """
+        super().__init__()
+        self.logger = logger
+        self.schema = schema
+        self.role = role
+
+    def _check_authorization(self, user: str, access: str):
+        """
+        Check whether an API call is able to access data, based on the API's
+        authorization header, the requested user, the requested access
+        policy, and the API's role.
+
+        If "user" is None, then the request is unauthenticated. READ operations
+        may be allowed, UPDATE and DELETE operations will not be allowed.
+
+        If "access" is None, READ will assume we're looking for access only to
+        public datasets.
+
+        for API_OPERATION.READ:
+
+            Any call, with or without an authenticated user token, can access
+            public data.
+
+            Any authenticated user can access their own private data.
+
+            Any authenticated ADMIN user can access any private data.
+
+        for API_OPERATION.UPDATE:
+
+            An authenticated user is required.
+
+            Any authenticated user can update their own data.
+
+            Any authenticated ADMIN user can update any data.
+
+        Args:
+            user: The user parameter to the API, or None
+            access: The access parameter to the API, or None
+
+        Raises:
+            UnauthorizedAccess The user isn't authorized for the requested
+                access.
+        """
+        authorized_user: User = Auth.token_auth.current_user()
```

Comment:
What does this do? I don't think I've ever seen this syntax before.

Reply:
Ah; it's a type hint on the assignment. I don't usually do that, but the Auth decorators are somewhat magical and this seemed like good documentation at the time. Not, of course, if the reader doesn't even recognize the syntax, though. Hmm.

Reply:
I'm sympathetic to the reluctance to make things harder for the reader, but I think that type hints ultimately will do the opposite, once we get our readers over the hump.  So, while I'm not going to call for using type hints anywhere outside of the function signatures, I'm disinclined to have them removed if they crop up in other places, especially if they can serve as a support to documentation.

Reply:
Ah, OK - thanks! Part of it was that I couldn't figure the correct incantation to ask Mr. Google. But I tried to check the grammar in the Language Reference and I couldn't make heads or tails of it: I don't think the grammar describes type hints in assignments (yet). But I agree that my ignorance should not stop forward progress.  Maybe a comment (here and in at least one more place  - so far I've only gotten to line 750 or so, so there might be more) to get the rest of us over the hump? I'll go read up on type hints now (I mean tomorrow...)

Reply:
The grammar is fine, my eyes not so much: now that I know what I'm looking for, I can find it. 20/20 hindsight is wonderful...